File: pyproject/baoli/MaxHeightOfTriangle.py
import math


class Solution:
    # 3200 三角形最大高度 数学法
    def maxHeightOfTriangle(self, red: int, blue: int) -> int:
        def maxH(r: int, b: int) -> int:
            # 首行放红球 1 + 3 + ... + (2n-1)
            n1 = math.floor(math.sqrt(r))
            h1 = 2 * n1 - 1
            # 2 + 4 + 6 + ... 2n => n + n²
            n2 = math.floor(math.sqrt(b + 0.25) - 0.5)
            h2 = 2 * n2
            return max(min(h1 + 1, h2), min(h1, h2 + 1))
        return max(maxH(red, blue), maxH(blue, red))
s = Solution()
print(s.maxHeightOfTriangle(2,4))

File: src/main/java/com/topic/baoli/MaxHeightOfTriangle.java
package com.topic.baoli;

/**
 * @author zc.zhou
 * @Description 3200 三角形的最大高度
 * @create 2024-10-15 9:08
 */
public class MaxHeightOfTriangle {
  public int maxHeightOfTriangle(int red, int blue) {
    return Math.max(maxHeight(new int[]{red, blue}), maxHeight(new int[]{blue, red}));
  }

  private int maxHeight(int[] balls) {
    int n = 1;
    int h = 0;
    int index = 0;
    while (balls[index]>= n) {
      balls[index] -= n;
      n++;
      index = (index + 1) % 2;
      h++;
    }
    return h;
  }
}
